pkg/component/controller/csrapprover.go

```diff
@@ -60,7 +61,7 @@ type CSRApprover struct {
 }
 
 // NewCSRApprover creates the CSRApprover component
-func NewCSRApprover(c *v1beta1.ClusterConfig, leaderElector LeaderElector, kubeClientFactory k8sutil.ClientFactoryInterface) *CSRApprover {
+func NewCSRApprover(c *v1beta1.ClusterConfig, leaderElector LeaderElector, kubeClientFactory k8sutil.ClientFactoryInterface) component.Component {
```

Comment:
What Is the reason for the change?

Reply:
The real interface that should be visible to the outside is just `component.Component`. The actual struct is just an implementation detail and can be hidden.

Reply:
It violates the "accept interfaces, return structs" rule. I'd prefer to keep it as is.

Reply:
Do you have a reference to this, so that I can read something about that?

I would argue that it's a common pattern to just expose interfaces (or traits, depending on the language) from factory methods. It makes it much clearer to understand what the return value should actually be used for. This is also not uncommon in Go, I'd say. All the standard library methods return plain errors, instead of e.g. *PathError or similar.

Reply:
This is a very common mojo of Go. I'm not sure who was the first one who said it. Probably, it [was](https://www.reddit.com/r/golang/comments/dfe1qr/who_first_said_accept_interfaces_and_return/) Dave Cheney's tweet, but he has deleted all his tweets. 

Google [gives](https://www.google.com/search?q=accept+interfaces%2C+return+structs&newwindow=1&sxsrf=ALiCzsYhz81ejp3U1mbmzgFhCe9ghRgaqQ%3A1651656940650&source=hp&ei=7EhyYpWyJPGSxc8P79-v4A0&iflsig=AJiK0e8AAAAAYnJW_Id2hpxIC9xCnq-NHHG77fYXSznn&ved=0ahUKEwiV2PDJxcX3AhVxSfEDHe_vC9wQ4dUDCAY&uact=5&oq=accept+interfaces%2C+return+structs&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEUABYAGDqA2gAcAB4AIABOIgBOJIBATGYAQCgAQKgAQE&sclient=gws-wiz) a lot of articles. 

> It makes it much clearer to understand what the return value should actually be used for.

Actually, it is the idea. Don't decide how to use it in the producer, let the consumer do it. The standard library has some exceptions for different reasons, but they are still exceptions.

Reply:
Thanks for your answer.

> Actually, it is the idea. Don't decide how to use it in the producer, let the consumer do it. The standard library has some exceptions for different reasons, but they are still exceptions.

That kinda contradicts how I think about types. While reading through all the people's comments on that topic, I clearly found voices who were saying that "preemptive" interfaces are just the wrong approach for Go, but not really giving any reason, besides that in Go, in contrast to some other languages, one can define interfaces after the fact on the consumer side. I don't buy that as is. Just because one can do it a certain way, it doesn't mean that that way is better than another one. I'd argue that interfaces neither belong to a consumer or a provider, but they literally stand in between of both of them. They're contracts, after all. So it makes perfect sense to me to think in interfaces first, to understand in how systems interact with each other, especially in more complex scenarios. Then I can start to look at actual implementations of both consumers and providers.

Go structs also have public interfaces, of course, but those are by far less easy to grasp from looking at the source code, as compared to interfaces. Go structs public methods are scattered across lengthy method implementations, where often enough public methods and private utility methods are mixed up, so there's no way to easily grasp the actual intent of the public API surface of a struct. One can address this when thinking about where to place which method in a source file, but most often, there are colliding interests: Either group the methods to better outline the public API, or group methods to better reflect the actual implementation. Usually the second one wins for good reasons.

Almost all of the component structs are just that: `Component`s. Some of them `ReconcileComponent`s. They need to implement those interfaces anyways, and, in fact, there are lot of interface checks in the code to ensure this. For me, it feels most natural to use those interface types in the factory methods. This communicates very clearly: No other interaction possible with the returned value. I don't have to check the code to find all public methods and check if there's anything outside the component interfaces that's used by k0s.

So, all in all, I prefer to return interfaces here for a single reason: Readability of intent. But that's of course just how I wrap my head around things. I am definitely on the idiomatic side of things and I now accept that, in this case, the idiomatic way of Go is not directly aligning with my train of thoughts :blush:

I'll change the return values to struct pointers again (also including the LeaderElector, for consistency :v:). It might make sense to make the structs private later on, though, since they are never used directly, just via the interfaces. Thanks again for pointing out that mantra. Definitely made me think a lot.